### ISSUE TITLE
a path towards moving from Allegro to SDL

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags#1096]
[**`ags/wiki/SDL-Migration-proposition.md`**](https://github.com/adventuregamestudio/ags/wiki/SDL-Migration-proposition)

The document is in the wiki so people can write in and clarify! If you are worried about losing your edit, do it using git, the document is `SDL-Migration-proposition.md` on the root directory.
- https: `git clone https://github.com/adventuregamestudio/ags.wiki.git`
- ssh: `git clone git@github.com:adventuregamestudio/ags.wiki.git`

---

I wrote above a [document](https://github.com/adventuregamestudio/ags/wiki/SDL-Migration-proposition) proposing a path toward migrating from Allegro, meaning, removing Allegro dependency completely, and using instead mostly SDL. It may turn out we need something else too.

Since last ags3-sdl2 port did not get merged, this tries to detail a plan, with **ags4** in mind and focus on complete removal of Allegro.

Please comment ~~, also accepting people to write the code~~. My central idea was to have code flowing towards this goal into AGS repo.

**Comment by ivan-mogilko** [comment]
I may be overthinking, but I don't like this sentence:
>  also accepting people to write the code. My central idea was to have code flowing towards this goal into AGS repo.

This is not the kind of task to be made by any multiple contributors, there must be a general plan and focused developer(s) to overwatch the process. It's better to have 1-2 people working on this to keep things organized.

**Comment by ericoporto** [comment]
I think a minimum of two people is required. I expressed badly, but writing code requires, other than the code: fix vcxproj files, fix makefiles, fix cmake files, testing and fixing initially in two (Windows and Linux?) , than three (MacOS), than 4 (Android), depending how libraries are inserted building or including these libraries (which means adding in the three dockerfiles), ... Even if is one for doing everything and the other to help testing. 

I think it's important to decide if we are building statically or dynamically (should game.exe accompany a SDL.dll ?) and how SDL gets in AGS project (code in repo or reference). It's important to note SDL may also go into editor due to Native.dll.

**Comment by JanetGilbert** [comment]
Hi, this is interesting to me as Wadjet Eye is attempting to get our games ported to Switch. We started working with Nick Sonneveld but then the project went on hiatus for a year and now we are working with Eduardo Garcia. We are using the method of using Nick Sonneveld's SDL version of AGS  combined with Ryan C. Gordon's port of SDL to Switch. I'm not sure how much we can share code as all the Switch stuff is under Nintendo's NDA so it can't be open source.

**Comment by ericoporto** [comment]
Hey @JanetGilbert , can you take a look at the path proposition and show to Eduardo Garcia too? The way the proposal diverges from @sonneveld approach is it prioritizes cutting ties from Allegro instead of relying in it's methods and reworking them in the backend. Sonneveld also changed other things that didn't make sense for me until you mentioned the Switch xD. I ommitted them here because I wanted the minimal things to move for SDL pure as fast as possible so other more interesting problems can be dealt with.

**Comment by JanetGilbert** [comment]
I'm not the techie on this project, AGS for SDL wouldn't compile for me! So Eduardo is handling the engine stuff, and I'm just going to be working on the scripting. I'm pretty sure it'd be easier to compile the less it depends on Allegro and associated libraries so I'm all for it. I've asked Eduardo to post here with his opinions.

**Comment by Arcnor** [comment]
Hi there. As @JanetGilbert mentions I'm working on porting AGS to the Switch. I started using Nick's port, although I've modified it a lot, which means my changes are far away both from AGS master & Nick's original changes :).

In any case, this is what I think it's best for a future version of AGS, in somewhat random order (sorry about that). I'm also assuming backwards compatibility is not desired, and if it is, separate tools can be written to deal with the migration. What I don't think should happen is maintaining support for both versions in the same engine, as there are certain things that cannot be reconciled (just my humble opinion, like everything I'm going to add here, please keep that in mind 😄 ):

First of all, in general (and stuff I'm already doing) should be dropping support for the following:
* Non 32 bit drawing code. If somebody wants 8 bit textures, they can use a palette they force themselves. If they want special effects, a plugin can be created that somewhat does those, maybe with shaders or maybe with some other magic, but maintaining the existing code for multiple color support is a real mess (IMHO, etc...)
* Any script function that modifies textures, like "Dynamic Drawing" or "Raw drawing" or stuff like that. This is due to performance, and not knowing if textures you are using have been modified or not, it's a headache down the road, specially depending on the kind of rendering you end up using.
* Any non-essential platform handling code, just use as much SDL (or similar libraries) as you can. Right now AGS deals with a million platform dependant things, and its abstracted in many ways to accommodate so. That made sense back in the day when multiplatform libraries were the exception rather than the norm, but nowadays there are plenty of options maintained by other people to help you, take advantage of that.

About the wiki entries themselves:
*Path handling*: Forget about using SDL or rolling your own, using `dirent` in Windows, etc, etc... There are many options in 2020 to deal with this as I mentioned before, and this is an excellent example of it. One of the best and most simple ones is just using PhysFS, which also happily works on consoles, and abstracts things like paths or even "containers" (you can have files inside a zip or in a directory, but in the end PhysFS abstracts all that). It's also by the main maintainer of SDL, so you know quality is at least as good.

*SDL Library placement in Filesystem* and *CMakeLists.txt, Makefile, Vcxproj and SDL in the CI and Dockerfiles*: This is a non issue, just use CMake for everything, SDL just works with it everywhere (I have it compiling for Linux, Windows, macOS and Switch on my own CI) and has its own CMakefile already. If you start juggling with different build systems you're going to have a bad day... Also, don't "ignore Android & iOS", it should work as well (I haven't added a build step for them, but I'm not expecting big issues with them, except maybe iOS for a bit, but it's a one time thing)

*SDL Video and Window handling and graphics* and *SDL and Directx9 renderer*: If you use the SDL Renderer API everything should just work everywhere (including consoles and smartphones), the only reason not to do it is in case you need special shaders or similar.
The only shader I've seen cannot be replicated directly is lighting, but that might just mean there might be another way of getting the same or a similar result using the SDL Renderer API instead of having the exact 100% result that AGS3 does today, given that AGS4 is a clean break.
It's also a good opportunity to drop the madness around transparency working in different ways depending if you have a value of 0, 1 to 250 or >250... Same for tinting, even if it doesn't match the AGS3 algorithm, just doing a simple color multiplication like the SDL Renderer API does should be enough...

All of that said, this is *not* the approach I'm going for, because the way in which AGS does rendering today doesn't match modern architectures (direct framebuffer access is not a thing anymore, anywhere) AND I need to keep compatibility with AGS3 because of Wadjet Eye's games, but the way I'm using is an even more extreme change so you might not want to do that for now, although I think it's the best option moving forward. As a summary, I'm not using SDL or its API for rendering, and also there is a lot of preprocessing of the data before it can be used with my fork of the engine, but it speeds up things tremendously, which is required for the Switch and maybe other platforms. I'll post more details if you're interested, although it's a work in progress and I'm a week or two away until I verify that everything works 100% like before.

But again, if you can use SDL Rendering and its limitations, it will work EVERYWHERE, which is work you don't need to do, as other people with way more experience is taking care of all the platform stuff for us.

*Reduce Allegro BITMAPS part 1*: Most of the mentioned libraries for antialiasing can just be dropped if the AA process is left to the driver (OpenGL, Metal, etc), but I guess you know this already, it's just not clear from the text.

For the rest I don't have any strong opinions yet, it's also a bit vague in many places like audio, so that's it for now.

**Comment by morganwillcock** [comment]
@Arcnor thank you for posting this here

> the way I'm using is an even more extreme change so you might not want to do that for now, although I think it's the best option moving forward. As a summary, I'm not using SDL or its API for rendering, and also there is a lot of preprocessing of the data before it can be used with my fork of the engine, but it speeds up things tremendously, which is required for the Switch and maybe other platforms. I'll post more details if you're interested

I think we also looking to change the project data formats and the whole game/media build pipeline, since what exists now is too dependent on the editor and hampers user workflow in general. It would be very interesting to learn more when it is convenient to you.

**Comment by Arcnor** [comment]
> @Arcnor thank you for posting this here
> 
> > the way I'm using is an even more extreme change so you might not want to do that for now, although I think it's the best option moving forward. As a summary, I'm not using SDL or its API for rendering, and also there is a lot of preprocessing of the data before it can be used with my fork of the engine, but it speeds up things tremendously, which is required for the Switch and maybe other platforms. I'll post more details if you're interested
> 
> I think we also looking to change the project data formats and the whole game/media build pipeline, since what exists now is too dependent on the editor and hampers user workflow in general. It would be very interesting to learn more when it is convenient to you.

You're welcome @morganwillcock :). In any case, it looks like I misinterpreted what this ticket was about and there was a discussion about it here https://discordapp.com/channels/221047797292597249/221048199895449600/725838953147662358 so I'll let others comment on the best way to continue.

**Comment by morganwillcock** [comment]
@Arcnor Unfortunately I missed the conversation in-person, but I did read it earlier and was a little disheartened by the direction it went in... I guess we will see how things turn out, but it would be good to stay in touch.

**Comment by ivan-mogilko** [comment]
From that discussion it appeared to me that we do not have the same understanding even of the goals, or relevant problems.

The above advices are good, even if we won't use all of them.

However, my impression was that this ticket is meant strictly for the purpose of replacing one backend library with another, which is a good first step. After that further changes could be done.

**Comment by morganwillcock** [comment]
I think the discussion was broader that what relates to this single ticket. I don't really see much logic to taking the branch that is decided to drop backwards compatibility and actually have significant modification, force yourself to port the graphics backend accommodating all the current features and issues, and then at the end of the process look at potentially removing some of those features and fixing some of those issues.

**Comment by ivan-mogilko** [comment]
Alright, then, maybe I don't understand the goals at all. Do I understand correctly that you suggest to first rewrite the engine's object logic and change backend only after?

**Comment by ivan-mogilko** [comment]
Also, does not this means that we should stop opening practical tickets like this one, and first actually write a plan with goals, and how to achieve these?

**Comment by ivan-mogilko** [comment]
> force yourself to port the graphics backend accommodating all the current features and issues, and then at the end of the process look at potentially removing some of those features and fixing some of those issues.

Another thing, tbh that's a bit abstract statement. The question is, what is ought to be removed, and whether porting introduces extra difficulties that may eventually be redundant.
From what I understood (but did not check closely yet) that port keeps existing AGS renderers, only using SDL for drawing bitmaps (same way as allegro is used now).

I think these concerns need to be clarified, to make sure everyone's talking about same thing.

**Comment by ericoporto** [comment]
https://github.com/adventuregamestudio/ags/compare/ags4...ericoporto:ags4--sdl2-sound-1

Above is a branch with Sonneveld's SDL2 sound implementation applied to AGS4 branch. 

Sound turned out to be the biggest impact on the branch.

It includes the following:
- SDL2;
- SDL_sound + MojoAL for audio playing and decoding;
- PhysFS (for the asset filesystem);
- AGS Strings based on `std::string`, with additional `nonstd::string_view` added for script dicts. These are apparently needed for the multithreading (may not be if we use mutexes?);
- C++ is kept at 11 but we can go 17 and just get the new features directly;
- AGS sound rewritten with a audiocore thread that keeps the information of sound in ags in sync with the one in the MojoAL/SDL_sound;
- only a single audio_clip class since everything goes through OpenAL compatible MojoAL instead of different decoders;
- modified Apeg for video (may be broken);

This branch has only been built using CMake 3.16 (in CLion 2020.1) on Ubuntu 20.04. For Windows VS2019 with CMake it will clash when building `allegro_main.cpp` since both Allegro and SDL2 add a macro that redefines the `main(argc,argv)` on Windows. For this I need someone with Windows experience to help.

---

Sorry for the confusion, this was a practical ticket of steps to get to SDL replacing Allegro. I didn't meant to diverge and discuss why.

**Comment by ivan-mogilko** [comment]
> AGS Strings based on std::string, with additional nonstd::string_view added for script dicts. These are apparently needed for the multithreading (may not be if we use mutexes?);

No, they are not needed for multithreading, and mutexes would be required to protect std::string as well, as they are not protecting themselves.

> C++ is kept at 11 but we can go 17 and just get the new features directly;

I would not do that right away unless critically needed (which I doubt it is). Every upgrade like this changes compiler requirement (and that's different on every platform) so this is something to be very careful about.

**Comment by ericoporto** [comment]
Yes, makes sense. This is the branch that has the code before the std::string switch and PhysFS addition: https://github.com/ericoporto/ags/commits/ags4--sdl2-sound

The game will play in silence if you build for release, but if one build for debug it should crash when trying to access a buffer before the data is there. 

Unfortunately I also only ran and built with CMake  through CLion 2020.1 and Ubuntu 20.04.

**Comment by ivan-mogilko** [comment]
Just a note to above branches, since I started investigating them, the commits there are real mess, with all due respect :-D: a single commit called "adds PhysFS support..." for example changes a whole lot of stuff at once, both related and not related to streams, moves code around etc. Need to try and split these in smaller changes to understand what's going on.

**Comment by ivan-mogilko** [comment]
As for the steps, I tend to agree that sound may be easiest to replace, because it's pretty much a separate thing.

I also learnt there is a way to use SDL in your program without SDL main and loop, which maybe will help to create intermediate builds, gradually replacing Allegro to SDL in various engine parts.

**Comment by ericoporto** [comment]
The commits are indeed a mess. Trying without extra libs is a good strategy! I am excited you are looking into this. :)

**Comment by ivan-mogilko** [comment]
I cloned the test branch and added few fixes to build on Windows: https://github.com/ivan-mogilko/ags-refactoring/tree/ags4--sdl2-sound-1

PhysFS appeared to be completely unnecessary, used in one class which may be disabled (there's few cases of unused new code overall, maybe prepared for the future).

I added SDL2 static lib from their website directly for this test. I don't know whether it's best option to have full SDL2 source in the repository, or why it was made so in the first place (for simplicity sake?).

**Comment by rofl0r** [comment]
> I don't know whether it's best option to have full SDL2 source in the repository

this doesn't sound good at all, much less so "best"

**Comment by ivan-mogilko** [comment]
@ericoporto unfortunately the code as it is does not work at least for me.

I suspect that sound playback init has some actions in a wrong order.

Without going too much into details, what happens there is that ALSource is getting polled by the internal thread (`mix_context` in mojoal.c) earlier than it's initialized with actual data (call to `alSourceQueueBuffers` in our audio_core.cpp). Because of that sound gets discarded by mojoal and never really plays.

I'll have to research this further, because I had no experience with this sound library before.

**Comment by ericoporto** [comment]
![image](https://user-images.githubusercontent.com/2244442/94315189-91291f00-ff71-11ea-9766-0517e8e99b4e.png)
@ivan-mogilko I get the following error with VS2019's CMake when building your branch.

I haven't build my branch on Windows, as I note there, the only place I tested at the time was Ubuntu 20.04 and latest CMake there. 

>I suspect that sound playback init has some actions in a wrong order.

This is possible, my first approach used only the audio core and nothing else (I think basically no changes except on the media folder). At the time I got a race condition trying to access the vox package.

this is the branch: https://github.com/ericoporto/ags/commits/ags4--sdl2-sound

this is the significant commit: https://github.com/ericoporto/ags/commit/611b12c021663591ca78aa94bd95bf071d0655d4

But after I pulled more things from Sonnevelds branch (the branch you looked into) then it worked.

**Comment by ivan-mogilko** [comment]
Ok, this is weird, but I think I found a bug in MojoAL source... (https://github.com/ivan-mogilko/ags-refactoring/commit/c23b596b91d600044583df729c41500ce06fc562#diff-baa934647843fd49d8f1863017135f30L4145)

The problem there was that `alGenBuffers` function that suppose to return fresh buffers to fill instead always returns first N buffer IDs regardless of whether they are currently free or not. Because of this our own list of buffers gets filled with same first 2 IDs, and engine tries to refill them while they still may be in use.....
Only after this change I can finally hear sound.

**Comment by ivan-mogilko** [comment]
Aha, apparently there *was* a bug, because newest library code has additional condition, compared to what we use:
https://hg.icculus.org/icculus/mojoAL/file/tip/mojoal.c#l4290

**Comment by ericoporto** [comment]
Weird question, but does it sound good? When listening, do you notice any stutter or something?

**Comment by ivan-mogilko** [comment]
This is a good question actually. I was testing one game ("Guard's Duty"), and on title screen I hear some noise on "strong" notes, like trembling. Original engine has this noise too, but for only a moment and then it goes away.

**Comment by ericoporto** [comment]
Yes, when I listened I felt it was different, like slightly louder, but not exactly. No problem with that, and not sure if something else influenced my perception at the time.

**Comment by ivan-mogilko** [comment]
After testing more, there's definitely something wrong, or unfinished in this branch, because not all sounds are playing. Maybe it cannot play multiple sounds simultaneously? For example in "Guard's Duty" intro sound fx are playing but background music is not.

Anyway, now when I see how it works in general, I'd like to make a step back and reapply only necessary code in clean way.

**Comment by ericoporto** [comment]
My commits are throwaway code. My reading regarding what Sonneveld's code was and was doing:
- simplify the clip types, due to one audio decoder library (either mojoal or OpenAL);
- the audio types are picked directly from the "file" extension of the audio in the asset package;
- the audio data is read from the asset package and passed to the decoder;
- the decoder will then pass to the actual sound sink in SDL sound (tbh I don't know why SDL2 is not used directly);
- it needs a way to communicate between the ags Engine, and the audio thread, so we can update audio positions on one way and also modify volume or panning in the other.  
- Audio position in current code is only ms and "implementation dependent ticks" are not implemented;

**Comment by ivan-mogilko** [comment]
@ericoporto I made a new wip branch here: https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--try-sdl2-sound

It has considerable less changes as I only pulled what was necessary to sound. Plus split decoder from audio_core.cpp because there was much code in it.

The library sources are still added to repo for simplicity sake, but we need to decide how to link them properly before actually merging this.

I did not add changes to CMake yet because I dont fully know how it works, so you'll have to do this if you want to build it with cmake.

**Comment by ericoporto** [comment]
Hey, I will test this! It's cool, I can add the CMake part! 

What do you think of a repository for all the source dependencies? The sources could be pulled from it either as a git submodule or through an event in the solution or a fetch script in the CMake.

**Comment by ivan-mogilko** [comment]
I pushed couple of more commits, mostly removing old allegro-based code, and a small fix.
The problem which I experienced with previous branch, where simultaneous sounds were not playing, is not reproduced in the new branch. Don't know what's the reason is, but I also used most recent MojoAL version, which may also affect the result.

I will test your game on master separately first.

**Comment by ericoporto** [comment]
Hey, the drawimage is fixed! With this working, I tested it with Future Flashback and the audio is perfect! Sometimes, in regular AGS 3.5.0, I had some audible stutters when saving game on Windows, but **this did not occur** when I tested the version on your sdl branch - I can see the frame briefly freezes but the music keeps going perfectly :D . I also did not notice any missing sound effects when playing. From using only the ears I could not pickup anything.

**Comment by ivan-mogilko** [comment]
Sound caching only matters for slow reading systems, it was made for PSP (which is no longer supported); it may be still useful for console ports maybe. Creating a new one might be easy, mostly a matter of finding a good spot in code to keep engine architecture clean.
Don't yet know what other points mean and whether they are important.

In terms of backward compatibility there are two main points of concern:
* whether all previously supported formats are playing;
* synchronizing sound with game.
Because sound was non-threaded and played roughly fixed amount of sound each game tick, number of games were syncing game events and animations with sound position. I have a very rough idea of how this behavior may be simulated with a threaded audio, but would rather leave such compatibility problems until after SDL is fully integrated.

I'd like to create a proper branch in the main repository, forking from current master (ags3), and merge sound port there to have something persistent.
What may still be necessary for this start?
* Make a good decision on library sources;
* Update makefile and CMake lists;
* Ensure this works at least on Windows and Linux for now.
* Config options? I made it have only 1 generic option to turn audio on/off in a game for now, but I wonder how exactly OpenAL chooses what drivers to play sounds with. Maybe it also has a configuration? Or this could be done later too.

**Comment by rofl0r** [comment]
how exactly OpenAL chooses what drivers to play sounds with. Maybe it also has a configuration?

http://web.archive.org/web/20130225015051/http://connect.creativelabs.com/openal/Documentation/OpenAL%201.1%20Specification.htm#_Toc199835899

`ALCdevice * alcOpenDevice (const ALCchar *deviceSpecifier);`

btw, there's also the excellent OpenALSoft implementation for linux.

may i ask why openal was chosen over SDL_sound, as the port was intended to use SDL?

> I'd like to create a proper branch in the main repository

please, in case you already have binary stuff checked into your temporary branch the commits adding them need to be amended to remove it, otherwise the binary blob will forever stay in the repo's blob graveyard and everybody will have to download it.

> Make a good decision on library sources;

should definitely stay outside the AGS repo, otherwise it will bloat up like a balloon.
it would be preferable to have a git subproject, so people not needing the library sources (like linux users that have the libs installed systemwide) won't have to check the subproject out.
but even better than this would be a buildscript for windows users that just downloads the sources from another repo hosted in the ags organisation.

**Comment by ivan-mogilko** [comment]
> may i ask why openal was chosen over SDL_sound, as the port was intended to use SDL?

Both are used in this port variant, and MojoAL is a OpenAL implementation based on SDL.
I dont know much about this frankly, and still must take a closer look, but from my current understanding, SDL_sound is used to read and decode file format, while MojoAL provides sound mixer and playback of raw wave data.

> please, in case you already have binary stuff checked into your temporary branch the commits adding them need to be amended to remove it,

Yes, I marked these commits as [TEMP], and keep noting that these are to make preliminary testing faster.

**Comment by rofl0r** [comment]
> MojoAL is a OpenAL implementation based on SDL.

oh i see, it's an API on top of an API! layers upon layers. i wonder who had that idea...

**Comment by ivan-mogilko** [comment]
What's the difference between that and OpenALSoft?

**Comment by ericoporto** [comment]
OpenAL is just a specification. MojoAL is an implementation of the spe

We could move everything from `/libsrc` to a repository `ags-dependencies`. 

In the future if this turns out to be a pain - because of the many ags branches, we could just have things in. The idea of `ags-dependencies` would be being able to just get the source tar file from GH instead of caring for git cloning it. For CMake it could be fetch from there from CMake itself so it's part of the build process. I really would like source dependencies instead of binary ones - makes easier anytime it needs to be built for mobile or in the future, consoles.

I think SLN/Vcxproj also support fetching a source like this, like in a zip file, but I could not find anything on this, online? Also, not all libraries ship with vcxproj files, I think, so we would need to maintain the ones needed there - they should be fairly simple.

For commercially released games on Linux it's always better to pack all dependencies with the game, so it would be good to continue like this.

**Comment by ivan-mogilko** [comment]
From what I can see, OpenAL provides certain mixer and sound adjustment logic. So it depends on whether we need this layer of logic or not. If not, we could try using SDL sound directly, but then there's this concern: won't we end up writing similar logic ourselves?

**Comment by rofl0r** [comment]
> What's the difference between that and OpenALSoft for instance?

openal is a library interface that wraps OS-specific sound driver libraries and provides an uniform, platform-indepent API to play sound. for example, on linux it would probably use libasound (alsa) or OSS behind the scenes.

openalsoft is an implementation of that API which sits on top of libasound.

SDL2's own sound functions https://wiki.libsdl.org/CategoryAudio do the same thing as OpenAL, they too provide such an interface.

now if you use an openal wrapper that sits on top of SDL, you end up with 2 layers of indirection: openal -> libSDL2 -> libasound. needless to say that each indirection has a certain cost.

however, the SDL2 sound interface is quite rudimentary and uses a callback driven design which is rather unpleasant to use, (and requires the user to unpack the audio manually, e.g. by using libogg) that's why many games prefer to use sdl_mixer: http://www.libsdl.org/projects/SDL_mixer/ .

interestingly, the SDL2 audio documentation itself recommends to use either openal or sdl_mixer: https://wiki.libsdl.org/SDL_MixAudioFormat

so maybe that is the reason mojoal is in use here.

**Comment by rofl0r** [comment]
> I would rather not discuss that and move with the other things.

i think discussing it is fine, but i agree that since the sound stuff seems to work for the moment it would be more productive to continue with the port. whether direct use of libsdl's sound support is preferable can be evaluated when the rest works.

**Comment by rofl0r** [comment]
> We could move everything from /libsrc to a repository ags-dependencies

at the moment libsrc only contains a few scripts to download the source anyway

**Comment by sonneveld** [comment]
G'day,

Sorry I have had little time to contribute lately (and in the future) but just wanted to document what I remember from my SDL2 port branch. I'm heartened to see that you're considering my code!

I was regularly rebasing the SDL branch on top of the latest of AGS. And then I was rebasing on top of that for the port. So the branch wasn't really intended to be merged in verbatim but to be cherry picked as we migrated various systems to use SDL.

So yeah, sorry for the "mess"! I was using individual commits for features where I could but ultimately it complicated rebasing as there were more commits where I had to invidually fix conflicts. I always intended to break out those large commits into something easier to review before creating pull requests.

The various libraries were deliberately commited to make it easier for me to make various patches. That's why I added libraries in multiple commits, one for the original pristine library code and then patch commits on top of that. This made it a lot easier for development but I didn't really intend for these to always live in the ags repo (with the exception of the allegro library since it would just be a shell around SDL)

Speaking of the allegro library, most the changes there were removing any allegro apis that we didn't use and implementing SDL "drivers" that spoke to sdl. That's all contained in a single src/sdl2/system.c for easy access.

With sound, I needed to be able to mix multiple sources and support the same audio formats that AGS needed. SDL_Sound is used for decoding but for mixing there were a few options:

1) Create an allegro sound driver using SDL. I believe Edward Rudd did this in the past for humble bundle linux ports. This might be a good first step but it relies on allegro doing all the audio mixing which didn't help if we ultimately wanted to remove allegro altogether
2) Use SDL. It doesn't provide a mixer so we would have had to write our own which I wanted avoid since we wouldn't be able to take advantage of any hardware mixing.
3) Use sdl_mixer. I investigated this but unfortunately it only provides a single "music" channel for streaming. It didn't really gel with the AGS audio api which lets you play streaming on any channel. Also, this may have changed, but it didn't really seem to be in heavy development, so we'd be swapping one unmaintained api for another.
3) Use OpenAL. This was supported natively on a few different platforms, allowed streaming from multiple sources and did mixing so it did feel like a good fit. For the platforms that don't natively support OpenAL (namely Windows), OpenALSoft is a pretty good implementation but it is GPL so we'd have to be careful how we include it. Fortunately Ryan C Gordon has created MojoAL which is a complete OpenAL layer on top of SDL and zlib licensed. It was still in development and was missing a couple of features like getting absolute position of playing sources (which may be implemented now) but it was a good fit if people didn't want to install OpenALSoft or the ancient creative openal drivers.

In regards to the physfs changes. This is unncessary for an initial sdl port, I just wanted to play around with ZIP file support and it seemed to fit in nicely. You could overlay multiple zip files to create a single pseduo file system. And it does create a clean api for accessing files on multiple platforms.

APEG should work okay. At least it worked last time I tested it. Although in the future it might be better to replace with something like ffmpeg. Unfortunately I found the ffmpeg api tricky to work with, so that would definitely be a future project if you wanted to go in that direction

In regards to the general changes to AGS for SDL support. The largest change was to use openal so we could avoid relying on allegro for mixing. Most of the other changes were related to input and changes to the various internal gameloops to ensure SDL can process events while waiting for mouse clicks or whatever. Otherwise it would look like the process had hung. There are a few places that I marked with AGS_DELETE_FOR_3_6 ifdefs, and they were mainly the older dirtyrect drawing code. I didn't think they would be necessary in a world where you only use OpenGL or Direct3D for rendering. But up to you I guess!

Anyway, hopefully that clears up some of my decisions.. and good luck!

**Comment by ivan-mogilko** [comment]
@rofl0r if I understand correctly, since OpenAL is an interface, we may simply switch implementation if current does not work out for some reason?

@sonneveld Hello, I was wondering what happened to you. Unfortunately, it took me a while to deal with other less important things and return to this task too.
I found and marked several of your old threads with different works (streams, strings, asset loader for switch iirc, and so on) and probably will be using them as a reference.
Something due mentioning, I think we decided to try port without having allegro wrap over SDL. That might require more code refactor but hopefully is doable and will result in a cleaner code.

**Comment by rofl0r** [comment]
@rofl0r if I understand correctly, since OpenAL is an interface, we may simply switch implementation if current does not work out for some reason?

indeed. one could for example use it with OpenALSoft, or the old reference implementation of creative labs.

> it's just that I am concerned that it may be an overkill to bring whole ffmpeg purely to play videos

i suspect it should be possible to build ffmpeg only with support for the few videocodecs that are used by AGS (also in order to circumvent GPL'd codecs that may be built into libavcodec&friends). but that would probably involve yet another "fork".
might be preferable to use the system-wide installed libraries via dynamic linking.

**Comment by ericoporto** [comment]
@ivan-mogilko for ags3.6 I think I would keep allegro even if only for the bitmap operations, while having mouse, keyboard, and window and others on SDL2. The bitmap usage can be replaced by something else (with some considerable code, with Pixman or with more human API, Cairo graphics), but the interfaces will require a lot of rewrite, and I think it may be easier to do in ags4 with potential feature dropping (like 16 bit colors).

**Comment by rofl0r** [comment]
> The bitmap usage can be replaced by something else (with some considerable code, with Pixman or with more human API, Cairo graphics)

could you provide an overview of required functionality that you think is lacking in SDL2 ?

**Comment by ericoporto** [comment]
I could not figure out how to write Text or draw a line (or a circle, ...). But you are right, SDL_Surface with SDL pixels should get most things going, it just will require a lot of rewrite (even our wrapper classes have Allegro expectations), which is why I am afraid of working on it in 3.6 - we can get benefit from SDL interfaces quickly going, like multimonitor support, high pooling mouse, reliable keyboard handling, the sound, ...

**Comment by sonneveld** [comment]
@ivan-mogilko 

> I found and marked several of your old threads with different works (streams, strings, asset loader for switch iirc, and so on) and probably will be using them as a reference (if that's okay legal-wise).

I give you permission to use any code from my ags repo forks. Considering that the exact commits won't be against my name, could I request that I be mentioned in the Credits as the author of the original SDL port code?

> Something due mentioning, we were discussing trying to port without having allegro wrap over SDL. Could you elaborate, because I forgot, what were the reasons you went with allegro wrap, was that for simplicity sake, or something more serious?

My first attempts at porting the engine to SDL removed the allegro library altogether. However I found that because we relied heavily on the different allegro blending methods, as well as different drawing routines (line drawing etc), I had to copy and paste a lot of that code verbatim. It also meant that the engine didn't have full functionality until I had ported all the different allegro methods to use SDL.

For reference, you can see that attempt here https://github.com/sonneveld/ags/commits/ags3--sdl2--old--2016-10-06 (I think I have an earlier branch from 2012 somewhere too)

The later attempts kept allegro in the loop. By implementing the SDL support as drivers, it reduced the surface area of the allegro api that I needed to port. I wanted to commit this cut down allegro layer first, to ensure that compatibility remained before removing allegro altogether.

Either way, we would still need to reimplement the **blending** and **drawing** implementations (or drop them if you're dropping compatibliity). Early on I believe we needed the allegro **fixed point maths** library for the path finding code but that has been rewritten. Also **colour conversion** code (i.e. 32bit rendering at 8 bit and in the other direction)

> it's just that I am concerned that it may be an overkill to bring whole ffmpeg purely to play videos (while we still are using SDL_sound etc for sound).

That's fair. There is definitely an overlap when it comes to audio codecs. However SDL_Sound has a much less restrictive license, so we can include that while leaving the GPL licensed ffmpeg as an optional feature.

In regards to using thirdparty libraries in general, it's all a trade off. AGS only has a small team of developers and I see the time would be better spent on the editor and implementing actual features for creating adventure games. Supporting multiple systems and audio/video codecs can be delegated to libraries. Game engines like godot have the luxury of extra developers which is why they can afford to ignore libraries like SDL and write drivers per OS.

@ericoporto 
> for ags3.6 I think I would keep allegro even if only for the bitmap operations, while having mouse, keyboard, and window and others on SDL2. 

Agreed. The bitmap operations are the main thing that might just need to be copy/pasted or reimplemented into the ags project.

**Comment by ivan-mogilko** [comment]
Alright so, presumably, traditional bliting/blending and primitive drawing operations are not covered by SDL to full extent? I will have to research this when I get to them. AFAIK allegro uses traditional memory format for bitmaps, so it should not be a problem to incorporate its functions withing sdl-based program...

Anyways, this kind of match the plan I was imagining, where first step is sound (most easy to replace), and next step is window + events loop.

**Comment by ivan-mogilko** [comment]
A tiny update, I started wip branch for step 2, which includes SDL window handling and input.
https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--try-sdl2-window
It's kind of rough so far, and I might post more details later. Unfortunately, there are some choices done by Nick in regards to key handling which seem bit weird, which may also mean that I dont realize some problem that he was trying to handle, so I will need more time to research this.

As of now:
* allegro is installed with "SYSTEM_NONE", which prevents it from creating now unnecessary window and grabbing input. Still need allegro to initialize, as all bitmap code and some utilities depend on it.
* old OpenGL and Direct3D renderers work fine in SDL window, **but** 1) need to update OpenGL renderer to create context through SDL - this way the code might become universal, without need to split between Windows/Linux etc; 2) software renderer must be replaced with one using SDL_Renderer for final present (Sonneveld's branch has one ready, so it's mostly matter of copying I hope).
* there's a silly mistake that the window appears only partially on screen, this is because I resize it after creation and forgot it also needs re-aligning on desktop.
* input does not really work yet, but at least event processing is going and you can drag window around and close it (no freezing).
* after latest WIP commit there's a strange issue with audio stopping playing after few seconds (but audio thread continues to run well).

**Comment by ivan-mogilko** [comment]
Remade https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--try-sdl2-window
* added fully working mouse code (or so it seems).
* left keyboard out for this time. There are few things that I still do not understand completely and wanted to read some more on how SDL handles it to be certain what the code does.
* fixed dissapearing music (trivial missing break in the case statement, and engine kept thinking that user alt+tabbed outside the window)

**Comment by rofl0r** [comment]
> if we completely remove exclusive fullscreen

what do you mean by that ?

**Comment by ivan-mogilko** [comment]
After an unfortunate delay, the branch is updated with working SDL key input.

I spent time mostly researching different approaches, but in the end decided to use Nick's code only partially and keep original simple virtual keycode value throughout the engine, only changing the way it's received - from buffered SDL events rather than polling allegro (from the look of the code, @sonneveld wanted to already begin replacing this old keycode with reading sdl data).

The main reason I did this is that IMO there has to be a larger engine redesign to make direct use of events convenient. I was also concerned that otherwise their presence may confuse other contributors.
Ideally, the event polling should pass events to some subscribers (engine components and game states, whatever you call them). Right now, because of how engine is structured, we have to save received event in a queue and then pass it "out" through a sequence of functions by demand.
Also, I am simply not certain what would be more convenient: to have SDL_event read everywhere, or invent some custom struct to pass only data we may need. But that's concern for next time.

---

> > if we completely remove exclusive fullscreen
> 
> what do you mean by that ?

There are three possible gfx modes: windowed, "desktop fullscreen" (borderless window covering whole desktop) and "exclusive fullscreen". I dont know if third one is supported on Linux oses, on Windows it's a part of DirectX functions, and supposedly can improve perfomance as all video resources are given to single task (as a downside switching away is slower as system has to unlock the mode first). People were telling me that "desktop fullscreen" becomes a better default on Windows too.

**Comment by sonneveld** [comment]
> @ivan-mogilko only changing the way it's received - from buffered SDL events rather than polling allegro 

If you mean removing a keyboard buffer altogether, I'm not sure that's the right way to go. The reason why I buffered the events is because we have code in `ags_getch` and `ags_clear_input_buffer` essentially wrapping allegro's `readkey`, which returned the next key from the keyboard buffer. (Reference https://liballeg.org/stabledocs/en/alleg006.html#readkey )

Simply polling will mean you might miss key events between frames.

**Comment by ivan-mogilko** [comment]
@sonneveld 
> If you mean removing a keyboard buffer altogether,

No, I keep the buffer mechanic of course, this is what I am saying in quoted sentence: I am replacing allegro poll with getting a key from buffered SDL events (that we buffer).

> oh another reason for keeping track of SDL events was because the gui text input needs actual ascii chars (upper and lower case chars) instead of scancodes. SDL doesn't have a mapping from SDL_KeyboardEvents to ascii.

Hmm, I could have missed something, I spent several days figuring out what your code means, and in the end decided to try an experiment, and found that it may be possible to map from SDL keydown event to ascii without SDL_TextInputEvent, by checking for all mod states, including SHIFT. This is why for the time being I added only keydown events handling, with plans to utilize SDL_TextInputEvent later for unicode support.
Basically, what I do is convert SDL keycode + mod combination into internal engine enum that merges both ascii and these special allegro-based values (Alt + xxx):
https://github.com/ivan-mogilko/ags-refactoring/commit/0c9ed8401f99426f8e7de7483c1a0f9952727919#diff-0213334fb1d5c42579e84c9f63ec0ec4b7480382fbcb164c4d661f1341691595R38

**Comment by sonneveld** [comment]
Ah fair enough.. I think I misunderstood.

Because it looked like the combined scancode/ascii was only used internally, I decided to bite the bullet and use the actual sdl scancodes where I could (that way I could use the SDL key defines everywhere internally). As far as I could tell, ags scripts only exposed the mapped scancodes, and not the combined ascii codes anywhere so you might need to be careful you don't expose that accidentally. Translated ascii is only used for gui and parsers so it made sense to make it understand SDL textinput instead.

**Comment by ivan-mogilko** [comment]
> edit2: Looking at your commit, does it deal with special characters? (e.g mapping shift-3 to # on us keyboards). This was the main reason why I decided not to convert scancodes to ascii characters myself.

Ah. No, it does not. For some reason I thought this will work because there are SDLK_ constants for these too and so these will be covered. I will investigate more then.

> It seemed much easier to use SDL events as much as possible, only converting when necessary

I tried that at first, but got worried that this will cause confusion to people, because whole rules for using these conversions is unclear. I hoped to do this after inventing some better looking way to handle these.

**Comment by sonneveld** [comment]
> For some reason I thought this will work because there are SDLK_ constants for these too and so these will be covered

I went through the same thought process! But I believe those SDLK constants are for keyboards that don't require a shift to produce those keys. You might see there isn't a one to one mapping between those and ASCII characters.

**Comment by ivan-mogilko** [comment]
@sonneveld ok, I restored SDL_TEXTINPUT handling: a7aafb82feb37dddec308a63fc891833ee8077f7

It seems generally working, will have to tidy up later maybe, and add mors comments to prevent same questions in the future.

**Comment by ivan-mogilko** [comment]
Updated branch with number of fixes, OpenGL renderer uses SDL code to be more universal, and software renderer rewritten to use SDL_Renderer.

Software renderer now works like this: evertyhing is drawn to virtual screen as usual, but then it's converted into SDL_Texture and put on screen by SDL_Renderer, which may be accelerated if supported. For example, on Windows it creates Direct3D (according to log).

There are certain differences from original code by @sonneveld, one notable is that I explicitly define the final rect where texture should be scaled to, thus allowing letterbox/sideborders modes. Also, I kept old fade-in/fade-out functions, mostly because of plans to be able to have a "debug overlay" over any game effects in the future. There is also a questions of proper renderer architecture with support for render layers of different resolution, and so on, so I try to keep changes minimal whenever possible for now.

The Hqx filters are removed completely, as these are slow and it's unknown if they are used by anyone anymore (esp. considering software renderer was working terribly slow on Linux for many people). In theory it might be possible to have them inserted before final texture scale, but that's better left for later consideration.

**Comment by ivan-mogilko** [comment]
Also, I found a bug in SDL :)
https://bugzilla.libsdl.org/show_bug.cgi?id=5329

**Comment by ivan-mogilko** [comment]
Rebased / cleaned branches:

Stage 1: sound
https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--sdl2-port1--sound
Stage 2: window, rendering and input (based on top of Stage 1)
https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--sdl2-port2--windowinput

@sonneveld I am adding following to some commits: `Based on the work of @sonneveld (Nick Sonneveld)`, hope this is right way to reference you.

**Comment by ivan-mogilko** [comment]
Made a quick test by removing allegro from the project, and following are things that AGS still uses from allegro:

* Bitmaps and software blitting & blending functions; the easiest is probably to copy necessary part of allegro source (that does not seem too much, but I only made an approximation so far).
* File system functions, such as "find files" and path utilities; I guess as an option we could use physfs for these if we plan on using physfs in the future in general.
* Unicode string functions and conversions, used literally in one or two places in the engine, but also used extensively in alfont.
* Alfont depends on bitmap and unicode allegro functions.
* Apeg is used to play theora videos, and depends on some allegro parts; bitmaps for certain, but I did not do full research yet.
* FLIC video reading is done through allegro.

**Comment by ericoporto** [comment]
Just a note about playing Theora Videos, a library appears to also exist from Icculus: https://icculus.org/theoraplay/

He also ships a video player using it with SDL2 here: https://hg.icculus.org/icculus/theoraplay/file/tip/test/sdltheoraplay.c

Unfortunately I couldn't find documentation on it or the library.

**Comment by ivan-mogilko** [comment]
Well, I pushed a first wip attempt of embedding a gfx part from allegro, it's a little dirty atm and I only included headers:
https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--sdl2-port3--removeallegro

**Comment by ivan-mogilko** [comment]
I checked TheoraPlay, but it seems to use its own thread types inside, which may not be consistent with our code.

Instead for now pushed @sonneveld 's changes to apeg, which make it use OpenAL. They seem to do the trick, only had to replace allegro timers with SDL's, and fix a mistake that made video play x2 faster than the sound...

**Comment by ivan-mogilko** [comment]
Hmm, so, I decided to not use physfs at the moment, as it turned out that it can only replace findfile functions, but not other path-related ones.

It looks like we will need to have a stripped allegro library in some form, with following contents:
* bitmaps, drawing, blending and palette functions
* functions for loading/saving bitmaps and pcx (latter is for that silly preload splash screen);
* playing FLIC;
* PACKFILE, stripped to only supporting user vtable (used only for FLIC iirc);
* fixed math functions;
* file path and find-file functions;
* unicode functions (used very sparingly, but maybe we can use these later in full capacity for utf8 support);

Perhaps we may have this stripped library in its own repository, optionally added to project as submodule. Then we may get rid of full allegro dependency already.

**Comment by rofl0r** [comment]
> It looks like we will need to have a stripped allegro library

do you have an estimate how big that will be ? if it's just a handful of files (like less than 3 KLOCish) it would IMO be a lot less hassle if it was in the main repo rather than a subrepo.

**Comment by sonneveld** [comment]
> It looks like we will need to have a stripped allegro library in some form

This fairly stripped down already https://github.com/adventuregamestudio/ags/tree/ags3--sdl2/libsrc/allegro  Includes most of the code you're talking about, with an sdl2 subsystem. The CMakefiles have been cut down for easy inclusion as well.

**Comment by ivan-mogilko** [comment]
> > It looks like we will need to have a stripped allegro library in some form
> 
> This fairly stripped down already https://github.com/adventuregamestudio/ags/tree/ags3--sdl2/libsrc/allegro Includes most of the code you're talking about, with an sdl2 subsystem. The CMakefiles have been cut down for easy inclusion as well.

Okay, I spent several days researching what could be cut off or replaced, but maybe I can reuse your library, except it may be cut even further imo. Primarily I hope things like drivers could be removed, but also things like video bitmaps (as we only need them as memory ones), and things like that.
I will check this out today.

**Comment by ivan-mogilko** [comment]
@sonneveld So, I pushed a temp change with your stripped allegro lib embedded in the engine,
https://github.com/ivan-mogilko/ags-refactoring/commits/ags3--sdl2-port3--removeallegro
and I continued removing stuff there, aiming at having no single "driver" or "system" object in the library, as imho our goal is to have only functions that work with data passed into them:
* timers, mutexes, threads etc,
* removed all bitmap types except "memory",
* removed gfx driver, screen bitmap and any mentions of those
Note I did not clean up the headers for unix etc platforms, but they may be done in same way as I did with the Windows header.

After all this, the noteable problem is FLIC playback. Unfortunately its API was not open enough and implementation was relying on own Allegro's timer (which I removed). So, instead I deleted playback functions from the library code and exported "speed" (basically fps) variable. My plan is to try and recreate playback function in the engine while only using Allegro to read FLIC frames. The function in question is quite simple (https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/src/fli.c#L845), and curiously Allegro seem to export necessary variables too (https://github.com/adventuregamestudio/lib-allegro/blob/allegro-4.4.3.1-agspatch/include/allegro/fli.h#L51).

Finally, there is a number of Windows-only variables that export DirectDraw and DirectInput interfaces. These were used for two things:
* playing AVI videos through Windows API (non portable);
* exposing these interfaces by plugin API (again, non portable);
These are mostly useless now when we have SDL backend, and I think we actually may just drop their support and maybe provide better alternative later if needed.

**Comment by rofl0r** [comment]
> I continued removing stuff there

i'd go about it the other way round: remove allegro completely, then copy into the tree the bare minimum the compiler complains about as missing.

**Comment by ivan-mogilko** [comment]
> i'd go about it the other way round: remove allegro completely, then copy into the tree the bare minimum the compiler complains about as missing.

I tried doing this at first, and it caused me lots of headache tbh.

**Comment by ivan-mogilko** [comment]
To amend above, in Allegro API things are arranged in files/sections pretty neatly, so cutting of might be easier as you know that you don't need this or that section.
I also do not want to cut literally every unused function, because there's still a chance some of these might come handy at some point, and it's just simplier to keep them.
Still there are maybe few more things that could be removed now, so might do that.

**Comment by ivan-mogilko** [comment]
Okay, I cautiously hope I'm done with this part.
https://github.com/ivan-mogilko/ags-refactoring/commits/ags3--sdl2-port3--removeallegro
* Removed even more code from allegro: soft 3d drawing mostly (I had a crazy idea that if we add simple 3D drawing API into AGS, then we might utilize those for software mode, but they may be returned later).
* Removed non-portable video playback based on DirectX from Windows version (it required directx interfaces provided by Allegro, but it's not clear if we could easily reimplement this since we use SDL2,.. and besides it's non-portable).
* Removed (or rather deprecated) non-portable parts in plugin API, which returned directdraw and directinput for the same reason as above. Frankly I don't remember any contemporary plugin that was using them.
* [Reimplemented FLIC playback function in the engine](https://github.com/ivan-mogilko/ags-refactoring/commit/502394ae96f781f9090e833d4a8b073d5eabbc15): it's roughly 20 lines of code, and still using allegro for reading FLIC frames.

There may still be some errors and compatibility issues, but guess we'll have to deal with them as they are found, after merging SDL into main branch.

---

So, what should be done now is:
* Finally decide on how to have all these library sources:
  - stripped allegro
  - SDL
  - SDL_Sound
  - MojoAL (couple of smaller files)
* Cleanup SDL branch, using final library locations
* Update CMake files, make sure it builds at least on Linux too
* PR, merge
* Update CI scripts (separate task)

**Comment by ericoporto** [comment]
About CMake files, I can do them and PR to where the branch is, making sure it builds on Linux. It would be a toned down version of the one here: https://github.com/ericoporto/ags

**Comment by rofl0r** [comment]
> Finally decide on how to have all these library sources:
>    stripped allegro
>    SDL
>    SDL_Sound
>    MojoAL (couple of smaller files)

my preference would be to add stripped allegro and MojoAL sources to the tree (MojoAL unfortunately has no releases, so one can't pin down a specific version easily), and "add" SDL* via a script that downloads release tarballs for windows builds, so only the script has to be checked in, just like we did for allegro in the past https://github.com/adventuregamestudio/ags/blob/master/libsrc/download.sh#L20

**Comment by ivan-mogilko** [comment]
Remade the branch: https://github.com/ivan-mogilko/ags-refactoring/tree/ags3--sdl2-port-full

Contains stripped allegro and mojoAL source, but requires SDL and SDL_sound to be added separately.

@ericoporto you mentioned that you can update CMakefiles, although I would prefer to add patches or cherrypick commits from somewhere instead of PR, because prs create merge commits.

**Comment by rofl0r** [comment]
> prs create merge commits

not necessarily, the github "merge PR" button has a dropdown option "rebase and merge" which omits the merge commit.
it can also be done manually by adding a remote for the other repo, checkout the PR branch, then rebase that branch on top of the branch it's supposed to be merged into, then checkout the base branch and run git merge pr-branch, which, when successful, shows "fast forward" which means no merge commit was created (optimally one tests this on a backup copy of the git repo in case something goes wrong).

**Comment by ericoporto** [comment]
Hey, @ivan-mogilko I did a bit of the work on building on Linux with CMake here https://github.com/ericoporto/ags/tree/ags3--sdl2-port-full 

I tested a bit with some games and it appears to work. There is a small catch for now that I don't know how to solve that [SDL_Sound adds a custom target on it's cmake and names it `uninstall`](https://hg.icculus.org/icculus/SDL_sound/file/9262f9205898/CMakeLists.txt#l208), and CMake targets should be unique. When I had it's sources in the repository, [my approach was to prefix it](https://github.com/ericoporto/ags/blob/ags3-emscripten/libsrc/SDL_sound/CMakeLists.txt#L243).I don't have an obvious solution just yet, so I mailed it's mailing list asking for help.

You can use my source files directly without referring to it's commits. If commits are intended to be ported over, I need to ask how they should be separated - I imagine one for CMake everything, and other for Linux build fixes, and one to up C++ from 11 to 14 in the Engine (for `make_unique`).

**Comment by rofl0r** [comment]
> and one to up C++ from 11 to 14 in the Engine (for make_unique)

if a C++-14 feature has slipped in, maybe one should try to find a backwards-compatible alternative instead of forcing all users to upgrade their compiler.

**Comment by ericoporto** [comment]
I won't have the C++ discussion with you @rofl0r , I literally wrote on my comment and it's 2020 now.

**Comment by rofl0r** [comment]
> upgrading is painless

it depends. if you have latest windows os and a MSDN subscription, then certainly it is easy to update your MSVC (if the huge download is not a problem with your internet connection and disk space). if you don't, then it might become a problem. it's also a problem if you use e.g. centos stable or ubuntu 16.04 LTS which doesn't provide a GCC with C++14.
also in case of MSVC, usually a compiler update also "updates" the list of supported target OS, for example it might remove support for windows XP, vista, or even 7.

**Comment by morganwillcock** [comment]
Unfortunately it might also raise the glibc requirement too, which might remove binary compatibility with older environments. So even being able to build it isn't necessarily enough.

**Comment by ericoporto** [comment]
I just tested in Ubuntu 16.04 and it builds and runs fine there. Here's GCC features per version: https://gcc.gnu.org/projects/cxx-status.html
CentOS case, apparently it has a way to target old glibc from newer gcc -> https://stackoverflow.com/questions/52609269/c-project-compiled-with-modern-compiler-but-linked-against-outdated-libstdc/52611144#52611144
In case of Windows I don't know.

> it might also raise the glibc requirement too

Sure. Our current solution is to build on a older Linux distribution (debian) and hope it will work on modern distros. Our [Dockerfile](https://github.com/adventuregamestudio/ags/blob/master/ci/linux/Dockerfile) doesn't specify which debian, but assuming Jessie, it has [gcc 4.9.2](https://packages.debian.org/pt-br/jessie/gcc), which support c++14 - according to above docs, I have not tested.

My polyfill is here because the single feature in use is really simple, in any case -> https://github.com/ericoporto/ags/blob/ags4--sdl2-sound-1/Common/util/nostd_make_unique.hpp

**Comment by morganwillcock** [comment]
> > it might also raise the glibc requirement too
> 
> Sure. Our current solution is to build on a older Linux distribution (debian) and hope it will work on modern distros. Our [Dockerfile](https://github.com/adventuregamestudio/ags/blob/master/ci/linux/Dockerfile) doesn't specify which debian, but assuming Jessie, it has [gcc 4.9.2](https://packages.debian.org/pt-br/jessie/gcc), which supports c++14 - according to above docs, I have not tested.

It is passed as an environment argument to the build system, so in theory it can build on multiple versions just by adding more container sources, but yes it is set to use Debian Jessie. From memory, I think we were quite lucky with the versioning for this one, in that the glibc version wasn't raised (so everything stayed working with the pre-built binaries that come with the editor) and it added a compiler feature that was needed for some builds.

**Comment by ivan-mogilko** [comment]
TBH I was quite lost why there's suddenly this issue of C++14 sprung out, as we currently do not use any C++14 in the code...
There was make_unique in one of the sonneveld's branches, but that piece of commit remained unincluded.

If it's desirable to move to the next C++ version we will have to first research where it is supported (and it's particular features, as new standards were sometimes not implemented fully at once), and make sure we don't loose some systems that are still in use or don't introduce difficult requirements for some of the port. We did this with C++11 in the past.

But ofcourse if it's just make_unique, then it would be much simplier to use our own implementation at the moment. I'd really prefer to not include additional dependencies while we are changing backend, as it's worrisome enough already :).

@rofl0r
> it depends. if you have latest windows os and a MSDN subscription
To clarify, we are now relying strictly on free (community edition) MSVC.
> also in case of MSVC, usually a compiler update also "updates" the list of supported target OS, for example it might remove support for windows XP, vista, or even 7.
I am not sure if it's connected with plain C++ support, it looks more like dependent on runtime library implementation. Since we moved to MSVS 2015 ags engine requires separate build configuration for Windows XP. I have not tried C++14 and onward on XP though. Windows 7 is still my os at the moment, and I wrote and run C++17 programs on MSVS 2019 without any problems.
This is only for the reference.

**Comment by ivan-mogilko** [comment]
So, I tried fixing linux compilation, using old Makefiles for now, made it compile but not link yet (wip commit):
https://github.com/ivan-mogilko/ags-refactoring/commit/718ebe6207ebe385eb6b6173b4b4fd800afd9762

There seem to be a problem with SDL_sound. Ubuntu only has old packages, iirc mine is 16.04 LTS, and it has only SDL_sound 1.2, based on SDL1, while our code requires SDL_sound 2.x. I can download it by hand, but not sure if it's an optimal way and how to include it in the makefile rather than requiring user to set variable externally prior to running make.

**Comment by ericoporto** [comment]
Icculus mercurial have links to download specific changesets too. Click on `files` at right of a changeset and then click on the format you want to download the files on top.
![Screenshot_20201116-143813_Chrome](https://user-images.githubusercontent.com/2244442/99287917-a60b8b80-2819-11eb-8849-b3816ab6fd41.jpg)
The download step could be either a download script or a step in the makefile - if downloading source and building is desired.

**Comment by rofl0r** [comment]
> while our code requires SDL_sound 2.x

isn't SDL2_sound already part of SDL2 ? i can only find SDL2_mixer as an external library.

**Comment by ivan-mogilko** [comment]
> isn't SDL2_sound already part of SDL2 ? i can only find SDL2_mixer as an external library.

No, it's a separate library: https://www.icculus.org/SDL_sound/

**Comment by rofl0r** [comment]
sigh. these clowns haven't published a release since they started work on SDL2 version, but one can download a tarball of latest mercurial as https://hg.icculus.org/icculus/SDL_sound/archive/9262f9205898.tar.bz2 (remove .tar.bz2 for windows unzip program)

**Comment by sonneveld** [comment]
> sigh. these clowns haven't published a release since they started work on SDL2 version

There's no need to be disrespectful to anyone here. Ryan C Gordon's providing his work for free and we're hoping to build on top of that work.

**Comment by ivan-mogilko** [comment]
I pushed last changes needed to build the engine on Linux (https://github.com/ivan-mogilko/ags-refactoring/commit/72f4bbe66bff70c4eb165733216d34eb3be7e51d).

There was another issue, the SDL version our sound code requires is at least 2.0.8, and one coming with Ubuntu 16.04 is 2.0.4, so I had to build necessary SDL version myself too along with SDL_sound.
I don't know if this is related, or I screwed up somewhere else, but when running engine I get errors from SDL where it cannot find display and audio device, so could not test this at runtime yet.

**Comment by ivan-mogilko** [comment]
@ericoporto tried applying your changes to cmake, but it does not work for me, and because I dont know cmake very well, cannot figure what errors mean.

**Comment by rofl0r** [comment]
> I pushed last changes needed to build the engine on Linux (ivan-mogilko@72f4bbe) 

i tried to compile this on linux, but i get 

```
In file included from ../libsrc/allegro/include/allegro/internal/alconfig.h:40:,
                 from ../libsrc/allegro/include/allegro/base.h:41,
                 from ../libsrc/allegro/include/allegro.h:25,
                 from libsrc/apeg-1.2.1/apeg.h:4,
                 from libsrc/apeg-1.2.1/adisplay.c:16:
../libsrc/allegro/include/allegro/platform/alucfg.h:34:39: fatal error: allegro/platform/alunixac.h: No such file or directory
 #include "allegro/platform/alunixac.h"
                                       ^
```
it seems to be still expecting system-wide installed allegro, even though it is now in-tree. probably adding the right `-Iinclude` to Engine/Makefile would fix that.

**Comment by ivan-mogilko** [comment]
Sorry, I missed that one "alunixac.h". It looks like that header has to be generated depending on system. It contains macros telling what system has enabled. Many of these definitions are no longer necessary for a stripped library, but some still do, like ones that tell endianess.

**Comment by ericoporto** [comment]
@ivan-mogilko my code should fail at building SDL_sound, the error should be it complaining that the target `uninstall` is a duplicate target when building it. The easy solution is simply provide an alternative CMakeLists.txt for it where this target is either removed, has a different name or it checks and only adds such target if it doesn't exist. I haven't done that yet because I thought I could get it fixed on upstream. I will add a commit giving a solution to this either tonight or tomorrow morning. (If you edit the fetched CMakeLists.txt for SDL_Sound and remove this target everything will build once CMake refreshes)

**Comment by rofl0r** [comment]
> Many of these definitions are no longer necessary for a stripped library, but some still do, like ones that tell endianess

i happen to have produced a header that should be able to tell endianness at compile time on almost all systems in the universe https://github.com/rofl0r/endianness.h

**Comment by ericoporto** [comment]
@ivan-mogilko  Added the changes to build on top of your test linux branch here : https://github.com/ericoporto/ags/commit/86ddc80acbbede931d4d1785fca6d6cf975308a0

My full branch with this is here: [**ericoporto/ags/tree/ags3--sdl2-port-full--testlinux**](https://github.com/ericoporto/ags/tree/ags3--sdl2-port-full--testlinux)

Can you test it and see if it works for you?

**Comment by ivan-mogilko** [comment]
Couple of more commits, I removed alunixac.h reference from alucfg.h and instead put some defines that are still in use in the stripped code there.

@ericoporto also applied your cmake update with some exceptions. Now it passes configuration step successfully, but every time I try doing cmake build something wrong happens during building SDL2, which crashes my linux vm (resets the guest os).

Could you clarify, why including extra <unistd.h> was necessary? I can build with existing Makefile well, in both debug and ndebug.

BTW, judging by cmake output, it still creates few allegro macros which we may deprecate, but I guess this is a minor issue.

**Comment by ericoporto** [comment]
unistd.h was added for the posix `chdir` and `getuid`, that header is available in anything not Windows as far as I know, so it should not be a problem. I believe it used to be automatically included in gcc but since gcc 5 it has to be done explicitly ([this link says 4.7](https://gcc.gnu.org/gcc-4.7/changes.html)), a similar case is reported [here](https://stackoverflow.com/questions/60127820/i-have-not-included-the-unistd-h-header-file-why-can-i-still-invoke-the-getpi).

The reported crash is very weird and I have no idea what it could be right now, can you give me some information about your VM so I can try to reproduce? It doesn't happen here but Linux is my primary OS on my devices :/ Can you tell me which GCC you are running in your Ubuntu 16.04?

Have you tried to clone and build my branch directly? Can you tell me how you are building? Here is some extra info on building with CMake: https://cliutils.gitlab.io/modern-cmake/chapters/intro/running.html

**Comment by ericoporto** [comment]
I changed the platform detection code to be done using preprocessor definitions in the C code instead of using CMake: https://github.com/ericoporto/ags/commit/dddc4cbe7b8c9059fc0aedaa87851ee89d25950a

This is the branch I made for this: [**ericoporto/ags/tree/ags3--sdl2-port-full--testlinux2**](https://github.com/ericoporto/ags/tree/ags3--sdl2-port-full--testlinux2)

**Comment by ivan-mogilko** [comment]
Could you explain why is this necessary, cannot we just pass necessary flag from makefile?
Also, strictly speaking MSVC != Windows, as there may be other compilers used on Windows.

**Comment by ivan-mogilko** [comment]
Regarding the previously reported crash,
* VM is Ubuntu 16.04 LTS
* VM RAM is 2 GB
* GCC is 5.4.0

I was building according to CMAKE.md in our repository.

I managed to build SDL 2.0.8 on my own previously. Should maybe try 2.0.12, as this is what cmake script is doing?

**Comment by ericoporto** [comment]
> Could you explain why is this alplatdetect is necessary, cannot we just pass necessary flag from makefile?

I decided to skip passing anything because I thought it was more acceptable since I got from previous messages  the CMake generated specific files weren't desirable. In this way it's just preprocessor code so it should work anywhere with any build system. 

I will focus on figuring out the crash for now.

**Comment by ivan-mogilko** [comment]
I think I found what's causing vm to crash, it's "--parallel" cmake option. It passes successfully without it... but then I got unexpected compiling errors in ags engine, so there's still something to fix...

**Comment by ivan-mogilko** [comment]
@ericoporto ok, I added platform auto-detection, hopefully it'll make it simplier. I can build through cmake now too.

**Comment by ericoporto** [comment]
Awesome! :D Great work!

**Comment by rofl0r** [comment]
> Couple of more commits, I removed alunixac.h reference from alucfg.h and instead put some defines that are still in use in the stripped code there (https://github.com/ivan-mogilko/ags-refactoring/commits/ags3--sdl2-port-full--testlinux).

thanks, it seems to build so far, apart from

../Engine/media/audio/openaldecoder.h:22:23: fatal error: SDL_sound.h: No such file or directory

this is really annoying, SDL_sound is a SDL1 library (latest release targets SDL1), then some work was done to make it compatible with SDL2, so there exist some versions (in git only) that are compatible to both SDL1 and SDL2, but one of the latest commits removed SDL1 support.
so distros that want to update this package need 1) get some untagged beta-quality git checkout, and 2) either drop support for SDL1, or somehow stuff SDL_sound master into a SDL2_sound package and hack library/include paths into the Makefiles of projects using it.
what would be needed here is that the authors finally make a release and rename it to SDL2_sound, just like all other 3rd party SDL2 libs did.

**Comment by ivan-mogilko** [comment]
@rofl0r I built and installed SDL_Sound myself from the latest mercurial commit using their cmake scripts, which ended SDL_sound.h in /user/local/include and SDL_sound.so (and accompanying files, *.a, *.so.1 etc) in /usr/local/lib.
Did not know any better so I used `-I /usr/local/include` as a "sdl_sound cflag" for now (and similar for lib).

@ericoporto what is your experience of running sdl port from this branch? I think I've broken something in package manager in my os during experiments (was trying to manually install newer SDL), and now SDL is acting up, only managed to run it in windowed mode and w/o any sound. Need to cleanup os first before continuing tests.

**Comment by ericoporto** [comment]
@ivan-mogilko I played some Wadjet Eye Windows titles for fun using it and they seem to work. Sound is fine, including the voice acting! I am using my Ubuntu 16.04 which is ran on the machine directly to verify this.
I played some other games and did not noticed anything out of ordinary in them. 
I had two problems with my own game (Future Flashback), but they are solvable:
- My mouse wheel middle click doesn't seem to register.
- I knew that my Joystick SDL2 plugin could potentially crash when the Engine was using SDL2 too ([it crashes with 3.5.0 and SDL2 DIGI](https://github.com/ericoporto/agsjoy/issues/9)) but this doesn't happen with your branch, instead it's conflict result makes it not able to run in fullscreen but it runs fine when windowed! Super weird, but is a fix that should happen on my end.

**Comment by rofl0r** [comment]
> ../Engine/media/audio/openaldecoder.h:22:23: fatal error: SDL_sound.h: No such file or directory
> this is really annoying

i meant that it's hard to reason how to properly install SDL(2)_sound systemwide in the described situation, not that it's failing to find the header when installed (it isn't). i will ask other distro maintainers how they deal with this specific oddball package.

> Did not know any better so I set -I /usr/local/include as a "sdl_sound cflag" in our makefile for now (and corresponding for lib).

no, that's a bogus change. one should never hardcode non-standard local include directories but add them to CFLAGS manually during the build, if so required, like `CPPFLAGS=-I/usr/local make`.

**Comment by rofl0r** [comment]
> i will ask other distro maintainers how they deal with this specific oddball package.

due to the quirkyness of this package, it seems almost no distros provide it so far. its author should really be contacted to get this fixed.
additionally the `make install` step is broken in hg tip for the header SDL_sound.h (fails with error).
i have found only one distro who ships it, archlinux: https://aur.archlinux.org/packages/sdl2_sound-hg/ where one commenter proposed a very wise fix for the header install:

```
sed 's|FILES SDL_sound.h DESTINATION include|FILES src/SDL_sound.h DESTINATION include/SDL2|' -i CMakeLists.txt
```
this makes it so SDL_sound.h is installed into $prefix/include/SDL2/SDL_sound.h, which fixes the conflict with SDL1-sound (it would overwrite its header, making SDL-sound for SDL1 unusable), additionally it makes it automatically work because -I$prefix/include/SDL2 is already added as a CFLAG when running sdl2-config --cflags. i would recommend other distros to adopt this approach. the libraries are fortunately already in the SDL2* namespace.

the commit message of when i added the package to my distro has more details:
https://codeberg.org/sabotage-linux/sabotage/commit/6645182d209373b0d8a87769d77ce10816576812

**Comment by rofl0r** [comment]
> Couple of more commits, I removed alunixac.h reference from alucfg.h and instead put some defines that are still in use in the stripped code there (https://github.com/ivan-mogilko/ags-refactoring/commits/ags3--sdl2-port-full--testlinux).

having fixed the SDL2-sound issue, i tried to build again, but the following happens during linking:

```
libsrc/apeg-1.2.1/ogg.o: In function `alvorbis_close':
/tmp/ags-refactoring/Engine/libsrc/apeg-1.2.1/ogg.c:55: undefined reference to `vorbis_block_clear'
...
libsrc/apeg-1.2.1/ogg.o: In function `altheora_close':
/tmp/ags-refactoring/Engine/libsrc/apeg-1.2.1/ogg.c:70: undefined reference to `theora_clear'
...
libsrc/apeg-1.2.1/ogg.o: In function `altheora_get_frame':
/tmp/ags-refactoring/Engine/libsrc/apeg-1.2.1/ogg.c:410: undefined reference to `ogg_stream_packetout'
...
```
it seems the "apeg-1.2.1" code (is it still needed ? supposedly sdl2-sound can play ogg etc) has a hard requirement on libogg, libtheora and libvorbis, but the required libraries are not added to LDFLAGS when linking the engine. (after installing those 3 libs, link succeeds after manually appending `-ltheora -logg -lvorbis` to the linker command line extracted from make V=1).

**Comment by ivan-mogilko** [comment]
> it seems the "apeg-1.2.1" code (is it still needed ? supposedly sdl2-sound can play ogg etc) has a hard requirement on libogg, libtheora and libvorbis, but the required libraries are not added to LDFLAGS when linking the engine. (after installing those 3 libs, link succeeds after manually appending `-ltheora -logg -lvorbis` to the linker command line extracted from make V=1).

Maybe I miss something, but there are following lines in current Makefile-defs.linux:
<pre>
LIBS += $(shell pkg-config --libs x11)
LIBS += $(shell pkg-config --libs ogg)
LIBS += $(shell pkg-config --libs vorbis)
LIBS += $(shell pkg-config --libs vorbisfile)
</pre>

**Comment by ivan-mogilko** [comment]
apeg is (and was afaik) only used to play theora videos.
The error is strange, not sure how these were linked before? and I don't get same error.

**Comment by rofl0r** [comment]
i suspect the cause is in Makefile-defs.linux
```
LIBS += $(shell pkg-config --libs ogg)
```
when ogg is not installed, this command will fail (interesting that it didnt cause build error) and not return the library names.

btw i suspect the x11 pkg-config line can be removed, as sdl2-config should automatically add all required libs.

**Comment by rofl0r** [comment]
btw i did some performance tests using --fps and starship quasar (old freeware version) and the old engine seems to be a lot faster. it runs usually at over 60fps with software and GL driver, while the SDL2 branch crawls at around 10 fps (with PC speed throttled to 800MHz via powersave CPU governor). could it be due to liballegro being built with more optimized CFLAGS when it is built as a whole? my package recipe for allegro uses these CFLAGS:
```
-fno-unwind-tables -fno-asynchronous-unwind-tables 
-Wa,--noexecstack -g 
-O3 -fstrength-reduce -fthread-jumps -fcse-follow-jumps -fcse-skip-blocks
-frerun-cse-after-loop -fexpensive-optimizations
-fforce-addr -fomit-frame-pointer -mtune=generic 
```

**Comment by ericoporto** [comment]
I have an initial branch here for Android: [**github.com/ericoporto/ags/tree/ags3--sdl2-port-full--android**](https://github.com/ericoporto/ags/tree/ags3--sdl2-port-full--android)

I am not sure if this is the next step or if it's getting the CI working in a branch in the repo/PR - to trigger Cirrus builds, so I am leaving at this.

**Comment by ivan-mogilko** [comment]
@rofl0r 
> i was confused because the engine outputs INFO: OpenGL shaders: ENABLED
INFO: Created renderer: opengl even when software driver is active. interestingly though, the old version runs at 60 fps even with software renderer.

With the SDL port "software renderer" is software only until it finishes virtual screen, but final present is done using SDL_Renderer and virtual screen (allegro bitmap) is converted to SDL_Texture for this. I'd investigate why is it causing that much of a slowdown, as regular OpenGL renderer also converts raw sprites to textures. Maybe the conversion or present code is suboptimal.

Also, maybe SDL can use other renderer types too (non accelerated ones), we'd only need to code a configuration for this.

**Comment by ivan-mogilko** [comment]
@ericoporto 
> I am not sure if this is the next step or if it's getting the CI working in a branch in the repo/PR - to trigger Cirrus builds

I'd like to have a branch inside main repository with the complete part of the work, to not let it dissapear if I am suddenly distracted. But before that I'd probably have to do what I mentioned in this comment: https://github.com/adventuregamestudio/ags/issues/1121#issuecomment-726453148

**Comment by ivan-mogilko** [comment]
I opened a PR: #1136

it's mostly same code, only reverted some unnecessary makefile changes.

**Comment by ericoporto** [comment]
Awesome, after that is merged we close this and we can create separate smaller issues for the future TO-DOs, the CI fix (I have started this one, but I need a bit more time to figure out some details), build instructions update and the new ports. 

The AGS3 Android SDL2 port may have a few gotchas due to previous reliance on configurations in the custom Allegro4 Android backend. So at least for this one a separate issue is useful. AGS4 Android can have cool nice to have new things. :D

**Comment by ericoporto** [comment]
Hey, I am still working on CI fixes. I was having trouble with MSVC but latest SDL_Sound from 2 days ago has fixed all woes I had with it, including my need to rewrite it's CMakeLists.txt.

I installed Windows for the first time in years to be able to work on the Dockerfile more quickly here now and plan to have at least Windows and MacOS working in the CI this weekend and submit a PR.

**Comment by ivan-mogilko** [comment]
> I was having trouble with MSVC but latest SDL_Sound from 2 days ago has fixed all woes I had with it, including my need to rewrite it's CMakeLists.txt.

Ah, darn, sorry, I completely forgot about that, I had to fix SDL_Sound code in two places to build it in MSVS, but did not tell about this.

**Comment by ericoporto** [comment]
Since #1148 has been opened for the remainder fixes, I will close this issue so details can be discussed there.
